### ISSUE TITLE
Tracking: Rework Notebook images

### ISSUE THREAD

**Issue by DavidSpek** [kubeflow/kubeflow#5575]
/kind feature

## Goals

- Provide sensible Jupyter and RStudio dockerfiles
- Host the container images produced by said dockerfiles on ECR

## Background

The current tensorflow-notebook-image has not been updated in over a year, this is predominantly due to there not being a clear owner until the notebook working group was formed. Furthermore, the current build process is extremely complicated, it uses over [340,000 lines of code](https://raw.githubusercontent.com/kubeflow/kubeflow/master/components/tensorflow-notebook-image/releaser/lib/v1.9.7/swagger.json) to build a single container image. Clearly this is not best practice, and sets a bad example for users to build their own images.

A long standing request from the community is to support notebooks other than Jupyter. To date, supporting RStudio has required an [embedded Apache server](https://github.com/thesuperzapper/kubeflow-images/blob/master/rstudio/Dockerfile) to rewrite URL paths as R-Studio always hosts links at the HTTP root.

## Proposal

To enable native RStudio support, we propose adding the following annotations to the Notebook CRD:
- `notebooks.kubeflow.org/http-rewrite-uri`: sets `http[0].rewrite.uri` in the Notebook’s Istio VirtualService
- `notebooks.kubeflow.org/http-headers-request-set`: sets `http[0].headers.request.set` in the Notebook’s Istio VirtualService

To help help users understand how to make a working RStudio/VSCode container image, we will provide the following new Dockerfiles (and host build versions of them them):

\* denotes default images for the spawner_ui_config.yaml

Server Type | Name | Base Image | Extends Base With | PR
-- | -- | -- | -- | --
N/A | notebook-server-base | ubuntu:20.04 | common linux utilities | https://github.com/kubeflow/kubeflow/pull/5622
RStudio | notebook-server-rstudio | notebook-server-base |   | https://github.com/kubeflow/kubeflow/pull/5624
RStudio | \*notebook-server-rstudio-tidyverse | notebook-server-rstudio | Tidyverse (from conda) | https://github.com/kubeflow/kubeflow/pull/5711
VSCode | notebook-server-codeserver | notebook-server-base |   | https://github.com/kubeflow/kubeflow/pull/5625
VSCode | \*notebook-server-codeserver-python | notebook-server-codeserver | Conda installed Python | https://github.com/kubeflow/kubeflow/pull/5700
Jupyter | notebook-server-jupyter | notebook-server-base | Adds Python and Jupyter | https://github.com/kubeflow/kubeflow/pull/5623
Jupyter | \*notebook-server-jupyter-scipy | notebook-server-jupyter | scientific python packages (see official jupyter image) | https://github.com/kubeflow/kubeflow/pull/5699
Jupyter | \*notebook-server-jupyter-pyspark | notebook-server-jupyter-scipy | pyspark3 | https://github.com/kubeflow/kubeflow/pull/5630 (outdated)
Jupyter (Pytorch) | notebook-server-jupyter-pytorch | notebook-server-jupyter | Adds torch, torchvision, torchaudio | https://github.com/kubeflow/kubeflow/pull/5626
Jupyter (Pytorch) | \*notebook-server-jupyter-pytorch-full | notebook-server-jupyter-pytorch | useful python packages | https://github.com/kubeflow/kubeflow/pull/5631
Jupyter (Pytorch) | notebook-server-jupyter-pytorch-cuda | notebook-server-jupyter | Adds torch, torchvision, torchaudio with CUDA | https://github.com/kubeflow/kubeflow/pull/5626
Jupyter (Pytorch) | \*notebook-server-jupyter-pytorch-cuda-full | notebook-server-jupyter-pytorch-cuda | useful python packages | https://github.com/kubeflow/kubeflow/pull/5631
Jupyter (Tensorflow) | notebook-server-jupyter-tensorflow | notebook-server-jupyter | Adds tensorflow | https://github.com/kubeflow/kubeflow/pull/5627
Jupyter (Tensorflow) | \*notebook-server-jupyter-tensorflow-full | notebook-server-jupyter-tensorflow | useful python packages | https://github.com/kubeflow/kubeflow/pull/5632
Jupyter (Tensorflow) | notebook-server-jupyter-tensorflow-cuda | notebook-server-jupyter | Adds tensorflow + CUDA | https://github.com/kubeflow/kubeflow/pull/5627
Jupyter (Tensorflow) | \*notebook-server-jupyter-tensorflow-cuda-full | notebook-server-jupyter-tensorflow-cuda | useful python packages | https://github.com/kubeflow/kubeflow/pull/5632

To enable the user to pick RStudio and VSCode, we will provide an “Server Type” selector, and based on the user’s selection a different list of images will be deployed, and the required annotations/headers will be set. This will result in the addition of two sections to the spawner_ui_config.yaml:
- `imageVSCode`
- `imageRStudio`

## Steps:

1. Create a PR which updates the notebook-controller with the proposed annotations: 
   - [x] https://github.com/kubeflow/kubeflow/pull/5660 
2. Create multiple PRs which add the proposed Dockerfiles + CI/CD to build them:
   - [x] (See table for links to PRs)
NOTE: we must merge the PRs in the order of their inheritance, so that ECR will have the upstream image
3. Delete the old notebook images Dockerfile:
   - [x] TBA: https://github.com/kubeflow/kubeflow/tree/master/components/tensorflow-notebook-image 
4. Create a PR which updates the Spawner UI and spawner_ui_config.yaml as described:
   - [x] https://github.com/kubeflow/kubeflow/pull/5646

/cc @kubeflow/wg-notebooks-leads

**Comment by jbottum** [comment]
priority p2

**Comment by DavidSpek** [comment]
To give an update on the current progress of reworking the Notebook (and other WebIDE) images. The table below shows the image name, it's size, the packages that are installed for the various package managers and a link to the image on Dockerhub for testing purposes. The flowchart shows how the images relate to each other, along with the size and some package information. 

| Image      | Size | File | Packages | Dockerhub |
| ----------- | ----------- | ----------- | ----------- | ----------- |
| webide-base | 381 MB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-8f96b5653ae574488a22135cc94a0a8d803bc07ec66a30f15185a042ee22a6f9) | <details><summary>APT</summary><br>nodejs, yarn</details> | `davidspek/kubeflow-ubuntu-webide-base:0.2` |
| jupyter-base | 1.19 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-82960a78f358fd43487a5524b7c75606cc56ee067dd59d6b38fedaf445cdef9d) | <details><summary>APT</summary><br>bash, bzip2, ca-certificates, curl, git, gnupg, locales, lsb-release, software-properties-common, sudo, tzdata, unzip, vim, wget, zip, kubectl</details> <details><summary>Other</summary><br>Miniforge3, conda 4.9.2, Tini v0.19.0</details> <details><summary>Conda</summary><br>Python 3.8.6, pip</details> <details><summary>Pip</summary><br>notebook 6.2.0, jupyterlab 3.0.5, ruamel_yaml 0.16.12, kubeflow-fairing 1.0.2, kfserving 0.4.1, kfp 1.0.4, kfp-server-api 1.0.4, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-10c56992d5a84d2d7883e4108e97d1280ffdf373f9d8436548df30fc90a8b299)</details> | `davidspek/kubeflow-ubuntu-jupyter-base:0.5` |
| pytorch-base | 1.8 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-b07cd3fc15bed0545b85e2e35498d5a23b5c1479f20e2477e90dde4f443075a4) | <details><summary>Pip</summary><br>torch 1.7.1+cpu, torchvision 0.8.2+cpu, torchaudio 0.7.2, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-c5a673fe88dc6e99d7725b76adb9407fe5ff2b3266de70b1ee03c7402406533f)</details> | `davidspek/kubeflow-ubuntu-pytorch-base:0.1` |
| pytorch-full | 2.95 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-d85ea562d87f664abcf3bc283bfc883547d76d99a64fb7f65da900e05061421a) | <details><summary>Pip</summary><br>ipywidgets 7.6.3, ipympl 0.6.3, cloudpickle 1.6.0, dill 0.3.3, bokeh 2.2.3, seaborn 0.11.1, scipy 1.6.0, scikit-learn 0.24.1, scikit-image 0.18.1, pandas 1.2.1, matplotlib 3.3.4, xgboost 1.3.3, torchelastic 0.2.1, fastai 2.2.5, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-b51be28fbb7207947e6515c0338b540039368f04289634846b0c61107c8a85df)</details> | `davidspek/kubeflow-ubuntu-pytorch-full:0.1` |
| pytorch-base-cuda | 4.11 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-ed925f33be88520d070cbaede04b35174f6bd448b4ff7764981dcb93982c843a) | <details><summary>Pip</summary><br>torch 1.7.1+cu110, torchvision 0.8.2+cu110, torchaudio 0.7.2, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-17c1aa6cae338eb7a8600eeae7cc954969238e3d3d562b951e727ac766b76875)</details> | `davidspek/kubeflow-ubuntu-pytorch-base:0.1-cuda` |
| pytorch-full-cuda | 5.26 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-217120383857d5b03ea6c0f2040c6c241ea12f134f33938ac3e3d9f44d8a9b50) | <details><summary>Pip</summary><br>ipywidgets 7.6.3, ipympl 0.6.3, cloudpickle 1.6.0, dill 0.3.3, bokeh 2.2.3, seaborn 0.11.1, scipy 1.6.0, scikit-learn 0.24.1, scikit-image 0.18.1, pandas 1.2.1, matplotlib 3.3.4, xgboost 1.3.3, torchelastic 0.2.1, fastai 2.2.5, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-b51be28fbb7207947e6515c0338b540039368f04289634846b0c61107c8a85df)</details> | `davidspek/kubeflow-ubuntu-pytorch-full:0.1-cuda` |
| tensorflow-base | 2.25 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-0e672f36b9b9806b922392df0e59c051b052f0568b29b536e869a90f2ccc23ac) | <details><summary>Pip</summary><br>tensorflow 2.4.1, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-f14b7af9474f4da6f287d29890536e08d1d71b2cd7bc8f18e62ca3b82e993ab6)</details> | `davidspek/kubeflow-ubuntu-tensorflow-base:0.1` |
| tensorflow-full | 3.21 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-50ecae543d3010ab316c167941774c42baa6ffb91b200f1ba9c5caf587b837df) | <details><summary>Pip</summary><br>ipywidgets 7.6.3, ipympl 0.6.3, cloudpickle 1.6.0, dill 0.3.3, bokeh 2.2.3, seaborn 0.11.1, scipy 1.6.0, scikit-learn 0.24.1, scikit-image 0.18.1, pandas 1.2.1, matplotlib 3.3.4, xgboost 1.3.3, keras 2.4.3, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-fa890d2ac2a231cf907764a75e1585df3b58ba7969c84c8f9df61eff8c9fe25a)</details> | `davidspek/kubeflow-ubuntu-tensorflow-full:0.1` |
| tensorflow-base-cuda | 7.83 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-8655f4e7927913ffc8e4b49474b529fd3c4a5eb1d9e4b8932ef20eb29077231f) | <details><summary>APT</summary><br>cuda-cudart-11-1, cuda-compat-11-1, libcublas--11.1, cuda-nvrtc-11.1, libcufft-11.1, libcurand-11.1, libcusolver-11.1, libcusparse-11.1, libcudnn8 8.0.5.39-1+cuda11.1, libfreetype6-dev, libhdf5-serial-dev, libzmq3-dev,  pkg-config, cm-super, libnvinfer7 7.2.2-1+cuda11.1, libnvinfer-plugin7 7.2.2-1+cuda11.1</details> <details><summary>Pip</summary><br>tensorflow-gpu 2.4.1, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-107096ddeb27aed13da3f404d15415641f872dc3e098de28e69a5e245a0800ff)</details> | `davidspek/kubeflow-ubuntu-tensorflow-base:0.1-cuda` |
| tensorflow-full-cuda | 8.78 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-db67efdfb64d91ba730f781de3e1a2ca1f5dbb0b90c8bf5c1047c99e89a77173) | <details><summary>Pip</summary><br>ipywidgets 7.6.3, ipympl 0.6.3, cloudpickle 1.6.0, dill 0.3.3, bokeh 2.2.3, seaborn 0.11.1, scipy 1.6.0, scikit-learn 0.24.1, scikit-image 0.18.1, pandas 1.2.1, matplotlib 3.3.4, xgboost 1.3.3, keras 2.4.3, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-fa890d2ac2a231cf907764a75e1585df3b58ba7969c84c8f9df61eff8c9fe25a)</details> | `davidspek/kubeflow-ubuntu-tensorflow-full:0.1-cuda` |
| datascience-base | 3 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-a8613072422b4a075d741f30d57ab1b4968817b429e44920b950b5f388ceb294) | <details><summary>Other</summary><br>Julia 1.5.3</details> <details><summary>Conda</summary><br>r-irkernel 1.1*, r-base 4.0.3</details> | `davidspek/kubeflow-ubuntu-datascience-base:0.2` |
| datascience-full | 5.47 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-e15dcb321c155a0bf68034bb170f2f10eb3fa3a4295ce174647d8b25199f4469) | <details><summary>Pip</summary><br>ipywidgets 7.6.3, ipympl 0.6.3, cloudpickle 1.6.0, dill 0.3.3, bokeh 2.2.3, seaborn 0.11.1, scipy 1.6.0, scikit-learn 0.24.1, scikit-image 0.18.1, pandas 1.2.1, matplotlib 3.3.43, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-78a3e3e0aef71625b8f61f235c6c517d24ceb35963c04d9b0943ad8cbf103f6d)</details> <details><summary>Conda</summary><br>r-caret 6.0*, r-crayon 1.3*, r-devtools 2.3*, r-forecast 8.13*, r-hexbin 1.28*, r-htmltools 0.5*, r-htmlwidgets 1.5*, r-nycflights13 1.0*, r-randomforest 4.6*, r-rcurl 1.98*, r-rmarkdown 2.6*, r-rsqlite 2.2*, r-shiny 1.5*, r-tidyverse 1.3*, rpy2 3.4*</details> | `davidspek/kubeflow-ubuntu-datascience-full:0.3` |
| spark-base | 1.94 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-b096131a86d9c4531827e4a1ea9814162385a1d6798719c4006458e75aefe451) | <details><summary>APT</summary><br>openjdk-11-jre-headless</details> <details><summary>Other</summary><br>spark-3.0.1-bin-hadoop3.2</details> <details><summary>Pip</summary><br>pyspark 3.0.1, pyarrow 3.0.0, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-c41d7e2996079191899071b52ed25edf32b287fdbc975b407b381376ef1c5988)</details> | `davidspek/kubeflow-ubuntu-spark-base:0.2` |
| code-server-experimental | 1.08 GB | [link](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-3d57bacc28687dfec7d7d8d6d799473f4dbd5263862e836fac3c31141c9b7db6) | <details><summary>APT</summary><br>jq, nano, net-tools, nodejs, yarn, python3-pip, apache2</details> <details><summary>Other</summary><br>[code-server latest](https://github.com/cdr/code-server/releases), [s6 overlay v2.2.0.1](https://github.com/just-containers/s6-overlay/releases/tag/v2.2.0.1)</details> <details><summary>Pip</summary><br>kfp 1.3.0, kfp-server-api 1.3.0, kubeflow-fairing 1.0.2, kfserving 0.4.1, [link to requirements.txt](https://github.com/kubeflow/kubeflow/pull/5582/files#diff-66b3ebdaf6a3b3a5c21c8af5ec45c3cb83ca87efd46d0856739ffc4aecfad181)</details> <details><summary>Extensions</summary><br>None yet</details> | `davidspek/kubeflow-ubuntu-code-server-experimental:0.4` |

![images-reworkv5](https://user-images.githubusercontent.com/28541758/107354834-f15ad080-6ace-11eb-8d33-15b2c430d985.png)

**Comment by kimwnasptd** [comment]
I can see a lot of opened PRs that introduce many frameworks alongside different Notebook Servers, so I'll comment in this issue instead of each opened PR.

As mentioned in the Community Meetings and in the weekly WG meets supporting different types of Notebook servers for 1.3 is best effort. So there won't be timely reviews until we ensure the manifests for the WG's artifacts and the CI/CD processes are in place for 1.3.

Lastly we haven't yet finalized the design doc around how we will handle the Notebook images, where will they live, which images there are etc. So until the design doc effort is finalized we can't move forward with reviewing the opened PRs. There should be more traction on the design doc around the Notebook images once we've dealt with the more pressing issues I mentioned above.

**Comment by DavidSpek** [comment]
While there are many PRs here, half of those require essentially no review as all they do is install a package with pip or conda. Also, while the design doc is not yet finalized, the PRs for the images do follow the intention of the design doc (at least how I interpreted the discussions) and the suggested changes I have made to it. The PRs that need the most attention will be those in stage 1, followed by those in stage 2. After that, the only PR that requires some attention is TensorFlow with CUDA due to needing to install CUDA. Me and @thesuperzapper are trying to coordinate going through these PRs so they would be ready for 1.3.

**Comment by DavidSpek** [comment]
To go along with these images, I have actually started work on an Image Builder web app, that will allow users to easily extend (these) images with whatever pip and conda packages they need. However, I do not think this will be ready for 1.3 (even though all that is needed is some very simple front end development.

**Comment by kimwnasptd** [comment]
Now that we have set up the CI/CD #5482  I can start allocating more cycles to reviewing this effort. I'll also be updating our [design doc](https://docs.google.com/document/d/14ve8inClzkhIILAIp2FodemRtoS2hqanX8jll2lmRUE/edit#) as we go along step by step. I'm not expecting us to have everything documented in that doc from the beginning, but lets use this issue to write down our process, assumptions and next steps. In parallel, as we progress, I'll try to keep our design doc up to date.

@DavidSpek I see a lot of opened PRs for different Notebook-servers/web-IDEs right now, which IIUC are all grouped under this umbrella issue. The way I'll be reviewing this effort is by going over the PRs based on the server they use [ Jupyter, R-Studio, VSCode etc ], starting from Jupyter. 

So the first thing I'll be looking at will be the Tensorflow CPU ones https://github.com/kubeflow/kubeflow/pull/5627 https://github.com/kubeflow/kubeflow/pull/5623, which we could use as our example Notebook image for running Jupyter with Kubeflow.

**Comment by thesuperzapper** [comment]
@kimwnasptd we are about to raise a new proposal (like in 30+ min) which simplifies this.

I expect we will just edit the text of this issue.

**Comment by thesuperzapper** [comment]
We will also refactor the PR's so dont review them yet.

**Comment by kimwnasptd** [comment]
Thanks for the heads up! I'll keep an eye on the issue then. Although most probably the order in which I'll look at the PRs will be as described above

**Comment by thesuperzapper** [comment]
@kimwnasptd can you please give us your thoughts on the updated proposal above (and specifically on the container names so we can get @PatrickXYS to create the required ECR folders)

**Comment by DavidSpek** [comment]
Maybe it is possible to have the images for the notebooks server live under `public.ecr.aws/j1r0q0g6/notebook-servers/***:tag`, similar to how kfserving has stuff under `public.ecr.aws/j1r0q0g6/kfserving/***:tag`. This would keep them a bit more organized given the amount of images that will be created.

**Comment by kimwnasptd** [comment]
Thanks for updating the proposal! First off I'm OK with creating upstream images for all the above, but with some changes on the names.

The goal of the following naming scheme is to just further align the proposal with our design doc and the [out of scope](https://docs.google.com/document/d/14ve8inClzkhIILAIp2FodemRtoS2hqanX8jll2lmRUE/edit#heading=h.47k606j3mc2v) section.
Since we don't want to be providing images for different frameworks and sets of packages we should only select one image for each server/web-ide and mark them as the "official" example Notebook images. Then we can include, and build/publish, all the other ones and mark them as `contrib` ones.

So the only changes to the naming scheme I would propose are:
1. Have `example-notebook-server-{jupyter,rstudio,vscode}` images [ I guess we could also append the underlying framework, i.e. `example-notebook-server-jupyter-tensorflow` ]
2. For the others to be `contrib-notebook-server-{jupyter,rstudio,vscode}-<any-suffix-you-want>`

cc @StefanoFioravanzo @elikatsis

**Comment by DavidSpek** [comment]
@kimwnasptd I'd like to suggest expanding the "official" images to the ones that don't include a `*` at the moment. Esentially, these are the base images for Jupyter, RStudio and VSCode, along with PyTorch and TensorFlow both with and without CUDA. I think this is necessary purely because TensorFlow with CUDA requires some some more steps to get working, and providing only TensorFlow CUDA without the CPU equivalent and the equivalent for PyTorch seems weird. Also, there the maintenance on the TensorFlow CPU and PytTorch base images should be essentially non-existent. 

Also, I thought it was decided that all the images should be seen as examples so I am not sure what the distinction is you are trying to make between `example` and `contrib`. It will also make the image names very long, which I think might start causing problems in the UI (or maybe even kubernetes, no idea if there is a limit for this).

**Comment by PatrickXYS** [comment]
To help unblock us, let's start from minimal, I can't see there's any blockers in getting started from minimal changes.

Haste makes waste, let's start with `example-notebook-server-jupyter-tensorflow`, but keep the code with enough extensibility. Later on, we can move forward with other IDEs and DL framework.

**Comment by DavidSpek** [comment]
@PatrickXYS That notebook would still require the `notebook-server-jupyter` and `notebook-server-base` images. I'm not sure if you've looked at the Dockerfiles, but (other than TensorFlow with CUDA), this is what that Dockerfile looks like:
```
FROM notebook-server-jupyter:<tag>

COPY requirements-cpu.txt .
RUN python3 -m pip install -r \
    requirements-cpu.txt --quiet --no-cache-dir \
    && rm -f requirements-cpu.txt
```
Contents of `requirements-cpu.txt`:
`tensorflow==2.4.1`

This is what almost all of the Jupyter Dockerfiles look like as well. There really is not much too them.

**Comment by DavidSpek** [comment]
@kimwnasptd I think you'd probably like to review https://github.com/kubeflow/kubeflow/pull/5646 given that you have done so much work on the JWA.

**Comment by DavidSpek** [comment]
@kimwnasptd I've thought about it a bit, and naming the images `example-*` and `contrib-*` makes absolutely no sense. 
First off, these are the images from the WG-Notebooks, not as contributions, so naming them as such makes no sense.
Secondly, the distinction between `example-*` and `contrib-*` also makes no sense, as it was agreed upon that users would be informed that the Dockerfiles and the resulting images that provided should be seen as examples. Ergo, they are all examples and thus this does not need to be in the name of the images on the registry. Why stop at `example-notebook-server-jupyter-tensorflow`, why not call it `example-notebook-server-jupyter-tensorflow-warning-not-intended-for-actual-use-beware-do-not-pull`?
Thirdly, I don't think the actual images on the registry themselves serve any use as an example, it is the Dockerfiles that are the actual example for the users. Thus, we may as well never actually build the images from the Dockerfiles or push them to a registry as they would serve no purpose.

Why call it Notebooks Working Group if there is no desire to actually create notebook images. At this point, I am seriously wondering if the current Notebooks Working Group should change its name to UI Working Group (or something else) and that a Notebooks Working Group is formed that is actually devoted to providing notebook images. As was discussed in the last Notebooks Working Group meeting, it is the responsibility of this working to provide notebook images. I think it sends a bad message too both users and the other working groups when the Notebooks Working Group doesn't want to commit to providing notebook images or the only images they provide are called `example-*` and any actual use of these images is strongly discouraged every step of the way. I would see this as the Notebooks Working Group not fulfilling their core duty, providing notebook images, which was the original reason for the working group. 

There were many statements in the design doc that directly contradicted each other that I highlighted and later removed by rewriting sections based on what was discussed and agreed upon in the meetings. However, most of these changes were removed and it is again full of contradictions and does not form a cohesive dialogue. I strongly get the impression that most of the Notebooks Working Group does not actually want to commit to putting effort into providing notebook images. If that is the case, I will happily write a proposal to create an Actual Notebooks Working Group which will truly focus on providing notebook images for Kubeflow.

**Comment by thesuperzapper** [comment]
@kimwnasptd two points:
1. Please treat this proposal as a __replacement__ to the previous [design doc](https://docs.google.com/document/d/14ve8inClzkhIILAIp2FodemRtoS2hqanX8jll2lmRUE/edit#heading=h.47k606j3mc2v)
2. I __strongly believe__ we should keep the names of the images as we have listed in the table
    * NOTE: we can create a README.md in the folder containing the images which makes it clear they are intended as an __example structure__
    * NOTE2: I am more than happy to take the brunt of maintaining these images going forwards (if we follow the proposed structure), so don't get worried about extra work if people start using them as non-examples.

## The default images in `spawner_ui_config.yaml` under this proposal will be:

#### RStudio
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-rstudio-tidyverse`

#### VSCode:
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-codeserver-python`

#### Jupyter
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-scipy`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pyspark` 
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pytorch-full`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pytorch-full-cuda`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-tensorflow-full`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-tensorflow-cuda-full`

**Comment by thesuperzapper** [comment]
@PatrickXYS 

I expect all "kubeflow-notebooks" images to live under: `public.ecr.aws/j1r0q0g6/notebooks/{NAME}:{TAG}`

I propose you __create the following ECR locations__ for us (as outlined above):

## Notebook Controller:
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-controller`

## Notebook Server:
__Base:__
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-base`

__RStudio:__
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-rstudio`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-rstudio-tidyverse`

__VSCode:__
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-codeserver`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-codeserver-python`

__Jupyter:__
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-scipy`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pyspark`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pytorch`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pytorch-full`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pytorch-cuda`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-pytorch-full-cuda`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-tensorflow`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-tensorflow-full`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-tensorflow-cuda`
* `public.ecr.aws/j1r0q0g6/notebooks/notebook-server-jupyter-tensorflow-cuda-full`

**Comment by thesuperzapper** [comment]
@kimwnasptd in the interest of getting these changes in for 1.3 (as I have already seen buzz about support for VSCode online)

Please know I will take responsibility for maintaining these images, so dont worry about extra work.

**Comment by thesuperzapper** [comment]
@PatrickXYS continuing from my previous comment.

__I propose we move all the ECR images for the "notebooks" Working Group under:__
`public.ecr.aws/j1r0q0g6/notebooks/{IMAGE_NAME}:{TAG}`

__This would be the same as kubeflow-serving, which uses:__
`public.ecr.aws/j1r0q0g6/kfserving/{IMAGE_NAME}:{TAG}`

__Right now we are at the root (which will be quite messy going forwards):__
`public.ecr.aws/j1r0q0g6/profile-controller`
`public.ecr.aws/j1r0q0g6/notebook-controller`
`...`
As seen here: https://gallery.ecr.aws/?searchTerm=j1r0q0g6

**Comment by thesuperzapper** [comment]
I created issue https://github.com/kubeflow/kubeflow/issues/5675 to track the provisioning of ECR images needed to unblock this (with some slight renaming).

**Comment by StefanoFioravanzo** [comment]
> Also, I thought it was decided that all the images should be seen as examples so I am not sure what the distinction is you are trying to make between example and contrib.

@DavidSpek  I'll try my best to answer this question and hopefully clear any confusion.

Our intention with the design doc was to set actionable goals and reasonable first expectations. For this reason, our priority is to only have these images act as examples of how to configure and run notebook servers on Kubeflow.

Kubeflow's mission is to allow people to run generic ML tools on top of Kubernetes, and this means that we should focus on providing the runtime context. See https://github.com/kubeflow/kubeflow/issues/1643#issuecomment-425170924 and https://github.com/kubeflow/community/issues/344#issuecomment-651936272 for some similar discussion that happened in the past.

So, for 1.3, the WG will take responsibility for supporting these base example images which provide the simplest working execution environment.
At the same time, we will provide a list of contrib images which contain some extra functionality. You can think about it as a way to "bootstrap" the `contrib` folder, so that anyone can follow along and build up from there. Our responsibility will be to provide documentation and guidance on how to work and contribute to the `contrib` images.

**Comment by DavidSpek** [comment]
@StefanoFioravanzo As I noted above `SDKs and integrations from inside the Notebooks (e.g. Git inside Notebooks)` are part of what the working group is supposed to provide. Also, this entire idea of splitting the images by `example` and `contrib` is all fine and nice, this was not ever actually specified in the design doc, nor discussed or agreed upon in the working group meetings. Furthermore, the images described here are the ones provided by the working group and therefore do not qualify as `contrib` images. In regards to `actionable goals and reasonable first expectations`, this seems like what you do before creating something, but the images were already all created (and tested to be working) at this point by me. I also have not seen any of the WG-Notebooks leads actually respond to the recent issues people were having with the current notebook images. Those people have now started using my new base images to solve their problems. 

As said above, @thesuperzapper (and I) will take responsibility for the images.

**Comment by thesuperzapper** [comment]
@StefanoFioravanzo @kimwnasptd

As @DavidSpek stated, we are more than happy to maintains these images, so there will be no extra work from your behalf.

The "contrib" idea, that makes very little sense, as these images are literally not contrib, they are examples
(NOTE: We will make it clear in the README.md of the `components/notebook-servers` folder that they are examples)

Given the above, I don't see any reason not to move forwards ASAP, especially since all the code is is ready to go. 

_PS:  We have __[publicly committed](https://github.com/kubeflow/kubeflow/blob/master/ROADMAP.md#kubeflow-13-features-target-release-march-2021)__ to supporting RStudio/VSCode in 1.3, and there is __nothing stopping us__ except this unproductive back-and-forwarth. Unless you are going __to propose and code__ another way to support RStudio/VSCode in __the next 4 days__, we are going to have to go with this proposal._

**Comment by elikatsis** [comment]
> Why call it Notebooks Working Group if there is no desire to actually create notebook images.

I'll add my thoughts as well, since this is a very fundamental matter. 

@DavidSpek  I understand you point of view, but I think this sentence is making some wrong assumptions. Notebooks WG does not aim to establish a generic central repository and maintenance for a wide variety of images. As @StefanoFioravanzo mentioned above, our main focus should be to provide example images that focus on setting up the runtime environments, so that end-users can build their own images.

The first step to achieve this is to standardize on how popular Notebook servers can be run on Kubeflow. Of course, after we are satisfied with our basic example images and documentation, we can discuss about having more sophisticated ML layers.
This effort should happen organically though and with well defined processes. We shouldn't try to do it all at once just a few days before the first release cut.

**Comment by kimwnasptd** [comment]
> As @DavidSpek stated, we are more than happy to maintain these images, so there will be no extra work from your behalf.

It’s not a matter of responsibility, but a matter of process and consensus. It's a matter of documenting what we want to do and what the user expectations should be. And this is why we have the design document.

> The "contrib" idea, that makes very little sense, as these images are literally not contrib, they are examples

The example notebook images, which is also the title of the design doc, are the ones that are meant to be used as examples on how people can configure/run the different servers. Not ones that expose different frameworks and sets of packages, as per the in/out of scope sections and the iterations we had over the past month.

But because we don't want to exclude the work put in the other images we want to name them `contrib-*` and pulish them as well. The `contrib-` prefix means that while an image does not yet completely align with what was established from our process/design-doc [ extra packages/frameworks ] we still want to hear user's feedback. Which is also a good next step for a migration process.

> Please treat this proposal as a replacement to the previous design doc

@thesuperzapper  I don't understand why you made this argument. We have been iterating on this design document at least a month now, so I can't go forth with a proposal that tosses out of the window all this effort just a few days before the release.

**Comment by DavidSpek** [comment]
@kimwnasptd 
> It’s not a matter of responsibility, but a matter of process and consensus. It's a matter of documenting what we want to do and what the user expectations should be. And this is why we have the design document.

You committed to providing these images in the WG-Notebooks charter, which your design doc is undermining. The design doc that you agreed on outside of the working group is therefore not valid and should not be taken into consideration in this discussion.

**Comment by thesuperzapper** [comment]
@kimwnasptd @elikatsis that design doc was not accepted (in fact we have no current method for accepting any proposal) 

@kimwnasptd can you please explain what you are trying to achieve by blocking this Proposal and is associated PRs?

**Comment by thesuperzapper** [comment]
@kimwnasptd I will begin reviewing the non-image PRs (as they are not controversial), and are needed to meet the agreed target of RStudio/VSCode support for 1.3:
1. Add annotations to notebook-controller (`notebooks.kubeflow.org/http-rewrite-uri`, `notebooks.kubeflow.org/http-headers-request-set`)
    * PR https://github.com/kubeflow/kubeflow/pull/5660
2. Update the jupyter WebUI to support setting those annotations
    * PR https://github.com/kubeflow/kubeflow/pull/5646

**Comment by kimwnasptd** [comment]
> @kimwnasptd @elikatsis that design doc was not accepted (in fact we have no current method for accepting any proposal)

Three of the WG's tech leads, @stefano, @elikatsis and I, have authored this document and we've been iterating on it over the last month. Although I understand the argument that we don't have yet established processes on accepting proposals. This is something I was planning to tackle after 1.3 by having our design docs in github, but this is a different discussion on its own.

> @kimwnasptd can you please explain what you are trying to achieve by blocking this Proposal and is associated PRs?

Nothing. I want us to have example notebook images as much as you do, but I want us to do this in a documented way, by establishing a process and by making it clear what the goals of the WG are around these images are.

> @kimwnasptd I will begin reviewing the non-image PRs (as they are not controversial), and are needed to meet the agreed target of RStudio/VSCode support for 1.3

Thank you! I was also planning to start reviewing them asap since this can be done in parallel.

**Comment by PatrickXYS** [comment]
From community side, there are a few points I can help add:

1. Closer release date is not a valid reason we have to agree upon a proposal / design doc in a rush, if we failed to achieve what WG said w.r.t 1.3 release, then we fail and has to let it fail in some degree, we can continue our work by doing minor release like 1.3.1.
2. I agree we should establish the standardization and move forward with a formal design doc / proposal, get WG's consensus, if we fail to have consensus, we need to establish mechanism such as voting to have majority tech leads agree upon the proposal (this is based on the fact that no single company employees occupy majority seats in the WG)

**Comment by thesuperzapper** [comment]
@PatrickXYS 

Re 1: in general I agree, but I strongly believe we can make this work (as there is very little actual work left, and the only thing we disagree on is the names of images)

Re 2: in the Notebook WG, 4 of 5 members are from the same company (not sure if other WGs have the same issue)

**Comment by PatrickXYS** [comment]
> Re 1: in general I agree, but I strongly believe we can make this work (as there is very little actual work left, and the only thing we disagree on is the names of images)

I think the common goal of us is to help WG notebook get to a better state.
All the comments around image naming itself, it's not an issue of "As long as we have one more folk agree upon the proposal", it's the issue of "As a whole group, we can't get a consensus in standardized way". 

Again, we all want to help notebook WG comes to a unified artifacts management/naming proposal, to get to that stage, can you comment in the [design doc](https://docs.google.com/document/d/14ve8inClzkhIILAIp2FodemRtoS2hqanX8jll2lmRUE/edit#heading=h.47k606j3mc2v) for the improvements that you think the doc need to be done?

If you think the doc is not acceptable from your side. Can you help bring up another design doc on your behalf?

**Comment by DavidSpek** [comment]
@PatrickXYS While I appreciate you trying to help get this solved, that design doc is not the way this will get solved. The moment this design doc was presented by @kimwnasptd I took the time to go through it thoroughly and I highlighted all the sections that were contradicting each other or that didn't make sense. Then, after a period of no response, it was brought up in one of the meetings and a consensus was formed what the actual intention and plans were that were described in the design doc. Directly following this meeting, I took the time to rewrite most of the design doc according to what was discussed and agreed upon in that meeting. After this, the design doc was not touched for a whole month. Then, when my comments were being resolved (while they actually weren't) and my changes were being removed, I instantly tried to start the dialogue and practically begged @kimwnasptd to chat (both on slack as well as in the document he was actively editing), but my request to actually discuss the design doc was not even acknowledged. After this session of removing my changes without so much as a discussion, what came in it's place were more statements that made no sense, and actually breach the charter of the working group. 

In the following Australia friendly WG meeting, where once again only @kimwnasptd and @thesuperzapper as leads of the WG were present (and myself), again a consensus was formed and we agreed on the approach that would be taken. Part of this agreement was that @thesuperzapper and I would create a short issue describing the needed changes that followed what was agreed upon in the meeting so that @kimwnasptd could quickly agree with the high level overview and we could continue. However, as you can see, the needed changes are once again being blocked the picture is painted that the goal of the working group surrounding these images was not decided upon. While in actual fact, the goal and how the intentions of the working group will be communicated to the public has been discussed and agreed upon multiple times in the working group meetings. 

Furthermore, I find the most worrying part of this entire discussion that the leverage being used to block this is caused by leads of the working group that have not taken part in these discussions in the meetings (as they haven't been to the meetings in a long time) or in the discussion I tried to start in the design doc.

Even disregarding all of the above. As I highlighted in my earlier comment, it is the responsibility of the Notebooks Working Group to provide these images how they are presented in this issue. Other than @thesuperzapper who is tried to fulfill his commitment and obligation to the working group and, by extent, the Kubeflow project, the other leads of the Notebooks Working Group are failing or actively blocking their commitment and obligations. With that said, I would also like to highlight the following statement in the working group governance document:
`Leads of a role SHOULD remove any other leads or roles that ... or are not fulfilling their documented responsibilities for more than 1 month.`

I would also like to highlight that the blog post for the 1.3 release from the Notebooks Working Group will be basically empty if this gets blocked any longer, as my PRs are the only actual new features that would be released as a part of this working group. Which I think will severely damage the reputation of the working group.

**Comment by kimwnasptd** [comment]
Summing up our plan around the Notebook images and the next steps to wrap up the remaining work around 1.3, based on our latest discussions.

We will move forward with the above images and names. At the same time we will need to update the docs to note that these images only aim to show people how to configure and run the Notebook servers. For the more technical stuff, like folder names and READMEs we can discuss in the PRs. Moving forward we will also need to define our story and assumptions around the images' frameworks and start including CI/CD tests for these assumptions.

So the next steps for 1.3 are:
1. Review and merge the needed PRs to support more Notebook servers https://github.com/kubeflow/kubeflow/pull/5660 https://github.com/kubeflow/kubeflow/pull/5646
2. Start merging the Notebook Dockerfile PRs. @thesuperzapper @DavidSpek  the order for this should be the one I see in the issue description above right?
3. Create a PR to Kubeflow's docs to note the context of the images and give a heads up to people about the new images
3. Delete the old Notebook images https://github.com/kubeflow/kubeflow/tree/master/components/tensorflow-notebook-image
3. Extend the CI/CD for building the images [both in post and presubmit like we do for the other components]
4. Extend the current config of the Jupyter web app to include the new images 
5. ...
6. Celebrate

At the same time I'll try to keep our design doc up to date with our latest progress on the technical decisions.

And with that said, I'm glad we've reached consensus and really looking forward to what we can do!

cc @StefanoFioravanzo @elikatsis @yanniszark @thesuperzapper @DavidSpek

**Comment by DavidSpek** [comment]
@kimwnasptd The order above is correct, there are still a few TBD, but those will not take long to make so I will do that once the open ones are merged. 
For point 3 (also regarding the READMEs) I'll also get started on a PR. 
For point 5, I have started including the CI/CD in the PRs for the notebook images. This way, once it is merged the image is instantly built and the resulting image artifact can be used downstream. 
Depending on the timeline of when https://github.com/kubeflow/kubeflow/pull/5646 gets merged compared to the images, point 6 can be included in https://github.com/kubeflow/kubeflow/pull/5646.

**Comment by StefanoFioravanzo** [comment]
Thanks @kimwnasptd for the summary! I'll help drafting and reviewing the docs part as well, @DavidSpek looking forwards to the PR. If you need more assistance in the reviews and other blocking issues, don't hesitate to ping me

**Comment by DavidSpek** [comment]
@kimwnasptd @StefanoFioravanzo Regarding point 6, I am almost finished pushing initial images of the existing PRs to ECR to test the CI/CD for the images. Once those are done I'll add them to the `spawner_ui_config.yaml` in https://github.com/kubeflow/kubeflow/pull/5646 so that the images on my personal dockerhub can be removed from that PR.

**Comment by kimwnasptd** [comment]
@thesuperzapper how is the review for the Notebook images going? Are there any issues?

 We will need to cut a release branch around Monday so we will need to start wrapping up. The Jupyter web app and Notebooks controller PRs #5646 #5660 are almost finished, so this should be the only remaining work.

**Comment by kimwnasptd** [comment]
@thesuperzapper @DavidSpek what is the state of this?

Could you also provide me a list of images that are merged and have a built image ready? I'd like to 
1. update the config of Jupyter web app to use the latest images
2. keep track of the images we will publish

**Comment by DavidSpek** [comment]
Only RStudio still needs to be done. Here is a list of the images for the `spawner_ui_config.yaml`. Here is a snippet for the latest images for Jupyter and VS-Code. 

```
  image:
    # The container Image for the user's Jupyter Notebook
    value: public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:master-78840844
    # The list of available standard container Images
    options:
      - public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-scipy:master-78840844
      - public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch-full:master-de7c3f00
      - public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch-cuda-full:master-de7c3f00
      - public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-tensorflow-full:master-a9821baa
      - public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-tensorflow-cuda-full:master-a9821baa
  imageVSCode:
    # The container Image for the user's VS-Code Server
    value: public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/codeserver-python:master-c99f886b
    # The list of available standard container Images
    options:
      - public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/codeserver-python:master-c99f886b
```

**Comment by kimwnasptd** [comment]
@thesuperzapper @DavidSpek could you provide a list of pending PRs?

> Only RStudio still needs to be done

Is there an opened PR for this? What's the effort required?

**Comment by DavidSpek** [comment]
PR's are https://github.com/kubeflow/kubeflow/pull/5624 and https://github.com/kubeflow/kubeflow/pull/5711 (this one is not much work). We were already quite a ways through the RStudio image but Mathew left for a dinner break.
